[MicroBench] Added a log_vml version of the signed log1p kernel

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64205
* #64033

The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to @bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

navahgar committed:
[MicroBench] Added a log_vml version of the signed log1p kernel

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench does not seem to much different in terms of perf. Here are the numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45403 ns        45402 ns        15916 GB/s=2.58488G/s
SignedLog1pBench/NNC/10/1467            42203 ns        42196 ns        16590 GB/s=2.78128G/s
SignedLog1pBench/NNCLogVml/10/1467      42463 ns        42456 ns        16461 GB/s=2.76425G/s
```

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench does not seem to much different in terms of perf. Here are the numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45403 ns        45402 ns        15916 GB/s=2.58488G/s
SignedLog1pBench/NNC/10/1467            42203 ns        42196 ns        16590 GB/s=2.78128G/s
SignedLog1pBench/NNCLogVml/10/1467      42463 ns        42456 ns        16461 GB/s=2.76425G/s
```

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers.

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           47273 ns        47267 ns        15122 GB/s=2.48289G/s
SignedLog1pBench/NNC/10/1467            39137 ns        39132 ns        18174 GB/s=2.99908G/s
SignedLog1pBench/NNCLogVml/10/1467      17909 ns        17907 ns        38359 GB/s=6.55377G/s
```

Thanks to @bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to @bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to @bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to @bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to @bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar commented on benchmarks/cpp/tensorexpr/bench_signed_log1p.cpp:
```diff
@@ -22,7 +23,7 @@ class SignedLog1pBench : public benchmark::Fixture {
 
   void TearDown(benchmark::State& state) override {
-    TORCH_CHECK(at::allclose(ref_, output_));
+    TORCH_CHECK(at::allclose(ref_, output_, 1e-3, 1e-3));
```
The test fails when `allclose` uses the default threshold as before. And that happens only when `log_vml` is enabled (I just confirmed that again).

We will definitely have to debug the numerics before enabling `log_vml` as the default.

bertmaher replied:
Hmm.  So I have a test for log_vml (I think it’s in test/cpp/tensorexpr) that sweeps large swaths of floating point numbers to check for accuracy.  My recollection was that it was either bit-accurate or within 1 ulp if VML (it depended on the platform; on avx512 machines vml uses a slightly different algorithm). We should do a similar check for log1p.  It would be unfortunate if log(1+x) (our implementation) is different from log1p(x)

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).

navahgar committed:
Update on "[MicroBench] Added a log_vml version of the signed log1p kernel"


The `log_vml` version of the micro-bench is over **2x** faster than the `log1p` version. Here are the perf numbers:

```
---------------------------------------------------------------------------------------------
Benchmark                                   Time             CPU   Iterations UserCounters...
---------------------------------------------------------------------------------------------
SignedLog1pBench/ATen/10/1467           45915 ns        45908 ns        14506 GB/s=2.5564G/s
SignedLog1pBench/NNC/10/1467            40469 ns        40466 ns        17367 GB/s=2.9002G/s
SignedLog1pBench/NNCLogVml/10/1467      19560 ns        19559 ns        35902 GB/s=6.00016G/s
```

Thanks to bertmaher for pointing this out. 

Differential Revision: [D30644716](https://our.internmc.facebook.com/intern/diff/D30644716)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30644716).
